File: modules/openredirect.py
"""Open Redirect detection module."""
import re
from urllib.parse import urlparse, parse_qs, urlencode
from typing import List, Dict, Optional, Set

from core.vuln_types import VT
import logging

logger = logging.getLogger(__name__)

_CANARY = "https://evil.erebus-scanner.internal"


class OpenRedirectFinding:
    def __init__(self, vuln_type, url, parameter, payload, evidence,
                 severity="Medium", confidence=0.95, exploitable=True):
        self.vuln_type   = vuln_type
        self.url         = url
        self.parameter   = parameter
        self.payload     = payload
        self.evidence    = evidence
        self.severity    = severity
        self.confidence  = confidence
        self.exploitable = exploitable
        self.category    = VT.OPEN_REDIRECT

    def to_dict(self):
        return {
            "type":        self.vuln_type,
            "url":         self.url,
            "parameter":   self.parameter,
            "payload":     self.payload,
            "evidence":    self.evidence,
            "severity":    self.severity,
            "confidence":  self.confidence,
            "exploitable": self.exploitable,
            "category":    self.category,
        }


class OpenRedirectModule:
    """
    Detect open redirect vulnerabilities including HTTP redirects,
    meta refresh, and JavaScript-based redirects.
    """

    _REDIRECT_PARAMS: frozenset = frozenset({
        'redirect', 'url', 'next', 'return', 'returnurl', 'goto',
        'dest', 'destination', 'ref', 'redir', 'redirect_uri',
        'continue', 'forward', 'callback', 'target', 'link',
        'location', 'redirect_to', 'to', 'back',
    })

    _CANARY_HOST = "evil.erebus-scanner.internal"

    _PAYLOADS: List[str] = [
        _CANARY,
        f"//{_CANARY_HOST}",
        f"/{_CANARY_HOST}",
        f"\\\\{_CANARY_HOST}",
        f"@{_CANARY_HOST}",
        f"%2F%2F{_CANARY_HOST}",
        f"https:/{_CANARY_HOST}",
        f"http://{_CANARY_HOST}",
    ]

    _META_REFRESH_RE = re.compile(
        r'<meta[^>]+http-equiv[^>]*refresh[^>]*content\s*=\s*["\'][^"\']*url\s*=\s*([^"\'>\s]+)',
        re.I,
    )
    _JS_REDIRECT_RE = re.compile(
        r'(?:window\.location|location\.href|location\.replace)\s*[=(]\s*["\']([^"\']+)["\']',
        re.I,
    )

    def __init__(self, http_engine, evasion_engine=None):
        self.http    = http_engine
        self.evasion = evasion_engine
        self._seen: Set[tuple] = set()

    def scan(self, url: str, method: str = "GET", data: Optional[Dict] = None) -> List[OpenRedirectFinding]:
        results: List[OpenRedirectFinding] = []

        parsed = urlparse(url)
        qs     = parse_qs(parsed.query, keep_blank_values=True)
        params = {k: v[0] for k, v in qs.items()}

        if method == "POST" and data:
            params.update(data)

        for param in params:
            if param.lower() not in self._REDIRECT_PARAMS:
                continue

            key = (url, param)
            if key in self._seen:
                continue
            self._seen.add(key)

            finding = self._probe(url, param, params)
            if finding:
                results.append(finding)

        return results

    def _probe(self, url: str, param: str, params: Dict) -> Optional[OpenRedirectFinding]:
        parsed = urlparse(url)
        base   = f"{parsed.scheme}://{parsed.netloc}{parsed.path}"

        for payload in self._PAYLOADS:
            try:
                test_params        = dict(params)
                test_params[param] = payload
                test_url           = f"{base}?{urlencode(test_params)}"

                resp = self.http.get(test_url, allow_redirects=False)
                if not resp:
                    continue

                # HTTP-level redirect
                if resp.status_code in (301, 302, 303, 307, 308):
                    loc = resp.headers.get("Location", "")
                    if self._CANARY_HOST in loc:
                        return OpenRedirectFinding(
                            vuln_type="Open Redirect",
                            url=url,
                            parameter=param,
                            payload=payload,
                            evidence=f"HTTP {resp.status_code} → Location: {loc}",
                            severity="Medium",
                            confidence=0.97,
                            exploitable=True,
                        )

                # Meta refresh in body
                m = self._META_REFRESH_RE.search(resp.text)
                if m and self._CANARY_HOST in m.group(1):
                    return OpenRedirectFinding(
                        vuln_type="Open Redirect (Meta Refresh)",
                        url=url,
                        parameter=param,
                        payload=payload,
                        evidence=f"<meta refresh> → {m.group(1)[:120]}",
                        severity="Medium",
                        confidence=0.90,
                        exploitable=True,
                    )

                # JavaScript redirect in body
                m = self._JS_REDIRECT_RE.search(resp.text)
                if m and self._CANARY_HOST in m.group(1):
                    return OpenRedirectFinding(
                        vuln_type="Open Redirect (JavaScript)",
                        url=url,
                        parameter=param,
                        payload=payload,
                        evidence=f"JS redirect → {m.group(1)[:120]}",
                        severity="Medium",
                        confidence=0.88,
                        exploitable=True,
                    )

            except Exception:
                pass

        return None

File: scanner/internal/modules/openredirect/openredirect.go
package openredirect

import (
	"context"
	"crypto/tls"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/erebus/scanner/internal/crawler"
	"github.com/erebus/scanner/internal/modules"
)

const attackerDomain = "erebus-redirect-test.invalid"
const attackerURL = "https://" + attackerDomain

var redirectParams = []string{
	"redirect", "redirect_uri", "redirect_url", "next", "url", "return",
	"returnUrl", "return_url", "goto", "destination", "dest", "target",
	"redir", "ref", "callback", "continue", "forward", "location",
	"successUrl", "failureUrl", "loginUrl", "logoutUrl", "r", "u",
	"link", "to", "out", "view", "dir", "show", "navigate", "path",
}

// bypassProbes are variants of attackerURL designed to evade naive allow-list filters.
var bypassProbes = []struct {
	value string
	label string
}{
	{attackerURL, "direct"},
	{"/" + attackerDomain, "root-relative"},
	{"//" + attackerDomain, "scheme-relative"},
	{"\\/" + attackerDomain, "backslash-slash"},
	{"\\\\" + attackerDomain, "double-backslash"},
	{"https:\\" + attackerDomain, "https backslash"},
	{"https://x@" + attackerDomain, "@ bypass"},
	{"https://" + attackerDomain + ".target.com", "subdomain confusion"},
	{"https://" + attackerDomain + "%0a.target.com", "newline in host"},
	{"%2F%2F" + attackerDomain, "double-encoded //"},
	{"%5C%5C" + attackerDomain, "encoded backslash"},
	{"java%0Ascript:alert(1)", "javascript: newline"},
	{"data:text/html,<script>location='https://" + attackerDomain + "'</script>", "data: URI redirect"},
	{" " + attackerURL, "leading space"},
	{"\t" + attackerURL, "leading tab"},
}

type Module struct {
	noFollow *http.Client
}

func New(noVerify bool) *Module {
	return &Module{
		noFollow: &http.Client{
			Timeout: 10 * time.Second,
			Transport: &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: noVerify},
			},
			CheckRedirect: func(*http.Request, []*http.Request) error {
				return http.ErrUseLastResponse
			},
		},
	}
}

func (m *Module) Name() string { return "openredirect" }

func (m *Module) Run(ctx context.Context, page crawler.Page) ([]modules.Finding, error) {
	var findings []modules.Finding

	// Test query parameters
	u, err := url.Parse(page.URL)
	if err == nil {
		tested := make(map[string]struct{})
		for key := range u.Query() {
			if ctx.Err() != nil {
				break
			}
			if _, ok := tested[key]; ok {
				continue
			}
			if isRedirectParam(key) {
				tested[key] = struct{}{}
				findings = append(findings, m.testParam(ctx, page.URL, key)...)
			}
		}
	}

	// Test form hidden fields and text fields
	for _, form := range page.Forms {
		if ctx.Err() != nil {
			break
		}
		for _, f := range form.Fields {
			if !isRedirectParam(f.Name) {
				continue
			}
			findings = append(findings, m.testFormField(ctx, form, f.Name)...)
		}
	}

	// Check page body for JS-based open redirect patterns
	findings = append(findings, m.scanBodyRedirect(page)...)

	return findings, nil
}

func (m *Module) testParam(ctx context.Context, pageURL, param string) []modules.Finding {
	var findings []modules.Finding

	for _, probe := range bypassProbes {
		if ctx.Err() != nil {
			break
		}
		u, err := url.Parse(pageURL)
		if err != nil {
			continue
		}
		q := u.Query()
		q.Set(param, probe.value)
		u.RawQuery = q.Encode()

		req, err := http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
		if err != nil {
			continue
		}
		resp, err := m.noFollow.Do(req)
		if err != nil {
			continue
		}
		resp.Body.Close()

		if f := m.checkResponse(resp, param, probe.value, probe.label, u.String()); f != nil {
			findings = append(findings, *f)
			break
		}
	}
	return findings
}

func (m *Module) testFormField(ctx context.Context, form crawler.Form, fieldName string) []modules.Finding {
	var findings []modules.Finding

	for _, probe := range bypassProbes[:6] { // limit form probes
		if ctx.Err() != nil {
			break
		}
		vals := make(url.Values)
		for _, f := range form.Fields {
			if f.Name == fieldName {
				vals.Set(f.Name, probe.value)
			} else if f.Value != "" {
				vals.Set(f.Name, f.Value)
			} else {
				vals.Set(f.Name, "test")
			}
		}

		var req *http.Request
		var err error
		method := strings.ToUpper(form.Method)
		if method == "" {
			method = "GET"
		}
		if method == "POST" {
			req, err = http.NewRequestWithContext(ctx, http.MethodPost, form.Action, strings.NewReader(vals.Encode()))
			if err != nil {
				continue
			}
			req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
		} else {
			u, err2 := url.Parse(form.Action)
			if err2 != nil {
				continue
			}
			u.RawQuery = vals.Encode()
			req, err = http.NewRequestWithContext(ctx, http.MethodGet, u.String(), nil)
			if err != nil {
				continue
			}
		}

		resp, err := m.noFollow.Do(req)
		if err != nil {
			continue
		}
		resp.Body.Close()

		if f := m.checkResponse(resp, fieldName, probe.value, probe.label, form.Action); f != nil {
			findings = append(findings, *f)
			break
		}
	}
	return findings
}

func (m *Module) checkResponse(resp *http.Response, param, value, label, reqURL string) *modules.Finding {
	if resp.StatusCode >= 300 && resp.StatusCode < 400 {
		loc := resp.Header.Get("Location")
		if strings.Contains(loc, attackerDomain) {
			sev := modules.Medium
			if label != "direct" {
				sev = modules.High // bypass technique is more severe
			}
			return &modules.Finding{
				Module:   "openredirect",
				Severity: sev,
				URL:      reqURL,
				Param:    param,
				Payload:  value,
				Evidence: fmt.Sprintf("HTTP %d → Location: %s (%s bypass)", resp.StatusCode, loc, label),
				Detail: fmt.Sprintf("Open redirect via parameter %q using %s: server redirects to attacker-controlled domain %q. "+
					"Exploitable for phishing, OAuth token theft, and post-auth redirect hijacking.", param, label, attackerDomain),
				CWE:         "CWE-601",
				CVSS:        6.1,
				CVSSVector:  "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N",
				Confidence:  modules.Confirmed,
				Remediation: "Validate redirect destinations against an allow-list of trusted domains; reject external URLs; use a token-based redirect system",
				Tags:        []string{"redirect", "open-redirect", "phishing"},
			}
		}
	}
	return nil
}

// scanBodyRedirect detects open redirect vulnerabilities that operate via JS (location.href, meta refresh)
// rather than HTTP 3xx headers — these are often missed by header-only checks.
func (m *Module) scanBodyRedirect(page crawler.Page) []modules.Finding {
	body := string(page.Body)
	u, err := url.Parse(page.URL)
	if err != nil {
		return nil
	}

	var findings []modules.Finding

	// Patterns that reflect a query param into a JS redirect or meta refresh
	jsRedirectPatterns := []string{
		"location.href", "location.replace", "location.assign",
		"window.location", "document.location",
	}
	metaPattern := `http-equiv="refresh"`

	for _, pattern := range jsRedirectPatterns {
		if strings.Contains(body, pattern) {
			// Check if a known redirect param value appears near the pattern
			for key, vs := range u.Query() {
				if !isRedirectParam(key) || len(vs) == 0 {
					continue
				}
				if strings.Contains(body, vs[0]) {
					findings = append(findings, modules.Finding{
						Module:      "openredirect",
						Severity:    modules.Medium,
						URL:         page.URL,
						Param:       key,
						Payload:     vs[0],
						Evidence:    fmt.Sprintf("Parameter value %q reflected near %q in page body — possible JS-based open redirect", vs[0], pattern),
						Detail:      fmt.Sprintf("Potential JS-based open redirect: parameter %q is reflected near a %s call — manual confirmation recommended", key, pattern),
						CWE:         "CWE-601",
						CVSS:        5.4,
						CVSSVector:  "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N",
						Confidence:  modules.Potential,
						Remediation: "Validate redirect destinations against an allow-list; avoid reflecting URL parameters directly into JavaScript location assignments",
						Tags:        []string{"redirect", "open-redirect", "js-redirect"},
					})
				}
			}
		}
	}

	if strings.Contains(strings.ToLower(body), metaPattern) {
		for key, vs := range u.Query() {
			if !isRedirectParam(key) || len(vs) == 0 {
				continue
			}
			if strings.Contains(body, vs[0]) {
				findings = append(findings, modules.Finding{
					Module:      "openredirect",
					Severity:    modules.Medium,
					URL:         page.URL,
					Param:       key,
					Payload:     vs[0],
					Evidence:    fmt.Sprintf("Parameter value %q reflected in meta refresh tag", vs[0]),
					Detail:      fmt.Sprintf("Potential meta-refresh open redirect: parameter %q is reflected in a meta http-equiv=refresh tag", key),
					CWE:         "CWE-601",
					CVSS:        5.4,
					CVSSVector:  "CVSS:3.1/AV:N/AC:L/PR:N/UI:R/S:C/C:L/I:L/A:N",
					Confidence:  modules.Potential,
					Remediation: "Validate redirect destinations against an allow-list; avoid reflecting URL parameters into meta refresh tags",
					Tags:        []string{"redirect", "open-redirect", "meta-refresh"},
				})
			}
		}
	}

	return findings
}

func isRedirectParam(name string) bool {
	lower := strings.ToLower(name)
	for _, rp := range redirectParams {
		if lower == rp || strings.Contains(lower, rp) {
			return true
		}
	}
	return false
}
